typescript/infra/scripts/check/check-warp-deploy.ts

```diff
@@ -1,29 +1,101 @@
 import chalk from 'chalk';
+import { Gauge, Registry } from 'prom-client';
 
 import { WarpRouteIds } from '../../config/warp.js';
+import { submitMetrics } from '../../src/utils/metrics.js';
 import { Modules } from '../agent-utils.js';
 
-import { check, getCheckArgs } from './check-utils.js';
+import {
+  getCheckWarpDeployArgs,
+  getGovernor,
+  logViolations,
+} from './check-utils.js';
 
-async function checkWarp() {
-  const argv = await getCheckArgs().argv;
+async function main() {
+  const { environment, asDeployer, chain, fork, context, pushMetrics } =
+    await getCheckWarpDeployArgs().argv;
+
+  const metricsRegister = new Registry();
+  const checkerViolationsGauge = new Gauge({
+    name: 'hyperlane_check_violations',
+    help: 'Checker violation',
+    registers: [metricsRegister],
+    labelNames: [
+      'module',
+      'warp_route_id',
+      'chain',
+      'remote',
+      'contract_name',
+      'type',
+      'sub_type',
+      'actual',
+      'expected',
+    ],
+  });
+  metricsRegister.registerMetric(checkerViolationsGauge);
 
   // TODO: consider retrying this if check throws an error
   for (const warpRouteId of Object.values(WarpRouteIds)) {
     console.log(`\nChecking warp route ${warpRouteId}...`);
+    const warpModule = Modules.WARP;
+
     try {
-      await check({
-        ...argv,
+      const governor = await getGovernor(
+        warpModule,
+        context,
+        environment,
+        asDeployer,
         warpRouteId,
-        module: Modules.WARP,
-      });
+        chain,
+        fork,
+      );
+
+      await governor.checker.check();
+
+      const violations: any = governor.checker.violations;
+      if (violations.length > 0) {
+        logViolations(violations);
+
+        if (pushMetrics) {
+          for (const violation of violations) {
+            checkerViolationsGauge
+              .labels({
+                module: warpModule,
+                warp_route_id: warpRouteId,
+                chain: violation.chain,
+                contract_name: violation.name,
+                type: violation.type,
+                actual: violation.actual,
+                expected: violation.expected,
+              })
+              .set(1);
+            console.log(
+              `Violation: ${violation.name} on ${violation.chain} with ${violation.actual} ${violation.type} ${violation.expected} pushed to metrics`,
+            );
+          }
+        }
+      } else {
+        console.info(chalk.green(`${warpModule} checker found no violations`));
+      }
+
+      if (pushMetrics) {
+        await submitMetrics(
+          metricsRegister,
+          `check-warp-deploy-${environment}`,
+          {
+            overwriteAllMetrics: true,
+          },
+        );
+      }
     } catch (e) {
```

Comment:
Wondering if we should have some way to know that an error has occurred. We had a similar problem with key funder - we want to know when errors occur, but don't want to stop the funding of other keys/chains if an error does occur

There iirc we'd keep track if any errors were caught and would use that to figure out what code to process.exit with. I think we have some metrics on pod errors available in prometheus that can be used for an alert in this case

Reply:
This is a good idea, as an aside, what do you think about retries? I've observed some errors due to RPCs

Reply:
What kind of errors have you seen?

Was hoping that SmartProvider is used under the hood and handles the retries at that level

Reply:
I think theres a error after the smart provider has exhausted all RPCs right, I saw this error once